src/dividat-driver/firmware/main.go

```diff
@@ -86,29 +94,29 @@ func Update(parentCtx context.Context, image io.Reader, deviceSerial *string, co
 			ctx, cancel := context.WithTimeout(parentCtx, 60*time.Second)
 			legacyDiscoveredAddr, err := discover("_sensoControl._tcp", deviceSerial, ctx)
 			cancel()
-			if err != nil {
+			if err == nil {
+				dfuHost = legacyDiscoveredAddr
+			} else if controllerHost != "" {
 				fmt.Printf("Could not discover update service, trying to fall back to previous discovery %s.\n", controllerHost)
 				dfuHost = controllerHost
 			} else {
-				dfuHost = legacyDiscoveredAddr
+				fail = fmt.Errorf("Could not find any Senso bootloader to transmit firmware to: %s", err)
+				return
 			}
 		} else {
 			dfuHost = discoveredAddr
 		}
-	} else {
-		dfuHost = *configuredAddr
 	}
 
-	// Wait briefly after discovery to ensure proper TFTP startup
-	time.Sleep(5 * time.Second)
-
-	// Transmit firmware via TFTP
-	err = putTFTP(dfuHost, tftpPort, image)
+	// 4: Transmit firmware via TFTP
+	time.Sleep(5 * time.Second) // Wait to ensure proper TFTP startup
+	err := putTFTP(dfuHost, tftpPort, image)
 	if err != nil {
 		fail = err
 		return
 	}
-	fmt.Println("✓ Firmware transmitted to Senso.")
+
+	fmt.Println("Success! Firmware transmitted to Senso.")
```

Comment:
At this point we could try to transmit to a Senso to which we failed to send a DFU command.
Adding some print statements and using a non-sensical address. I added some additional print statements to better see what the code is doing: 
```diff
diff --git a/src/dividat-driver/firmware/main.go b/src/dividat-driver/firmware/main.go
index aad158b..8ef3503 100644
--- a/src/dividat-driver/firmware/main.go
+++ b/src/dividat-driver/firmware/main.go
@@ -61,6 +61,7 @@ func Update(parentCtx context.Context, image io.Reader, deviceSerial *string, co
 		fmt.Printf("Using specified controller address '%s'.\n", controllerHost)
 	} else {
 		// Discover controller address via mDNS
+		fmt.Printf("No controller address specified. Trying to find Senso via mDNS.\n")
 		ctx, cancel := context.WithTimeout(parentCtx, 15*time.Second)
 		discoveredAddr, err := discover("_sensoControl._tcp", deviceSerial, ctx)
 		cancel()
@@ -68,11 +69,13 @@ func Update(parentCtx context.Context, image io.Reader, deviceSerial *string, co
 			fmt.Printf("Error: %s\n", err)
 		} else {
 			controllerHost = discoveredAddr
+		  fmt.Printf("Discovered Senso at: %s\n", controllerHost)
 		}
 	}
 
 	// 2: Switch the Senso to bootloader mode
 	if controllerHost != "" {
+    fmt.Printf("Sending DFU command to Senso at: %s\n", controllerHost)
 		err := sendDfuCommand(controllerHost, controllerPort)
 		if err != nil {
 			fmt.Printf("Could not send DFU command to Senso at %s: %s\n", controllerHost, err)
@@ -87,11 +90,14 @@ func Update(parentCtx context.Context, image io.Reader, deviceSerial *string, co
 		dfuHost = *configuredAddr
 	} else {
 		ctx, cancel := context.WithTimeout(parentCtx, 60*time.Second)
+    fmt.Printf("Trying to find Senso in DFU mode.\n", controllerHost)
 		discoveredAddr, err := discover("_sensoUpdate._udp", deviceSerial, ctx)
 		cancel()
 		if err != nil {
 			// Try to discover boot controller via legacy identifier
 			ctx, cancel := context.WithTimeout(parentCtx, 60*time.Second)
+      fmt.Printf("Could not find Senso in DFU mode.\n", controllerHost)
+      fmt.Printf("Trying to find Senso via legacy identifier.\n", controllerHost)
 			legacyDiscoveredAddr, err := discover("_sensoControl._tcp", deviceSerial, ctx)
 			cancel()
 			if err == nil {
@@ -105,13 +111,18 @@ func Update(parentCtx context.Context, image io.Reader, deviceSerial *string, co
 			}
 		} else {
 			dfuHost = discoveredAddr
+      fmt.Printf("Discovered Senso in DFU mode at: %s\n", dfuHost)
 		}
 	}
 
+  fmt.Printf("Waiting 5 seconds for TFTP startup.\n")
 	// 4: Transmit firmware via TFTP
 	time.Sleep(5 * time.Second) // Wait to ensure proper TFTP startup
+
+  fmt.Printf("Transferring image to Senso at: %s.\n", dfuHost)
 	err := putTFTP(dfuHost, tftpPort, image)
 	if err != nil {
+    fmt.Printf("Failed to transfer image.\n")
 		fail = err
 		return
 	}
```

Then when I update using:
```
./bin/dividat-driver update-firmware -i <image> -a "nonaddress"
```

I see:
```
Using specified controller address 'nonaddress'.
Sending DFU command to Senso at: nonaddress
Could not send DFU command to Senso at nonaddress: Could not dial connection to Senso controller at nonaddress:55567: dial tcp: lookup nonaddress: no such host
Waiting 5 seconds for TFTP startup.
Transferring image to Senso at: nonaddress.
Failed to transfer image.
Could not create tftp client: resolving address nonaddress:69: lookup nonaddress: no such host

Update failed. Try turning the Senso off and on, waiting for 30 seconds and then running this update tool again.
```

Reply:
> At this point we could try to transmit to a Senso to which we failed to send a DFU command.

Yes, that is part of the cockroach strategy. Try everything in order to recover and pick up from any point at which we previously might have gotten stuck.

In this case, if a Senso was sent to update/bootloader mode before, it is going to remain there (it seems even past a reboot), so it _is_ going to be in bootloader mode and wait for the TFTP connection, but it is not going to offer the connection to receive the DFU command.